### PR TITLE
Create and parse vendor extension attribute

### PR DESCRIPTION
**Pull request by mariomaz**
See [PPM-136](https://jira.prplfoundation.org/browse/PPM-136)

### CODE REVIEW

**Review by rmelotte — APPROVED** [review]

**Comment by rmelotte** [comment]
I added a commit to run MAP-4.10.4 again with the Qualcomm controller.
Since it's also failing on master (with the Mediatek controller) I don't expect it to pass, but we can at least manually check that we now correctly parse 1 fronthaul and 1 backhaul.

**Comment by rmelotte** [comment]
The certification test currently fails here: 
```
DEBUG 21:44:17:666 <1998507040> son_slave_thread.cpp[4039] --> Received AP_AUTOCONFIGURATION_WSC_MESSAGE
ERROR 21:44:17:667 <1998507040> m2.cpp[313] --> TLVF: getAttr<cWscAttrVendorExtension> failed
ERROR 21:44:17:668 <1998507040> m2.cpp[32] --> TLVF: Not all attributes present
INFO 21:44:17:668 <1998507040> son_slave_thread.cpp[4045] --> Not a valid M2 - Ignoring WSC CMDU
```

File: framework/tlvf/src/include/tlvf/WSC/configData.h [hunk 1]
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
...
+            if (eWscWfaVendorExtSubelement::MULTI_AP_IDENTIFIER == vendor_data[index++]) {
+                if (0x01 != vendor_data[index++]) {
+                    return default_value;
+                }
+
+                return static_cast<WSC::eWscVendorExtSubelementBssType>(vendor_data[index]);
+            }
+
+            // Skip this subelement
+            index += vendor_data[index];
```
Comment (rmelotte) [inline framework/tlvf/src/include/tlvf/WSC/configData.h]:
```suggestion
            index += vendor_data[index] + 1;
```

File: framework/tlvf/src/include/tlvf/WSC/configData.h [hunk 2]
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
         auto attr = getAttr<cWscAttrMac>();
         return attr ? attr->data() : sMacAddr();
     };
-    uint8_t bss_type() const { return getAttr<cWscVendorExtWfa>()->subelement_value(); };
+    eWscVendorExtSubelementBssType bss_type() const
+    {
+        eWscVendorExtSubelementBssType default_value = eWscVendorExtSubelementBssType::TEARDOWN;
```
Comment (arnout) [inline framework/tlvf/src/include/tlvf/WSC/configData.h]:
```suggestion
        constexpr eWscVendorExtSubelementBssType default_value = eWscVendorExtSubelementBssType::TEARDOWN;
```

File: framework/tlvf/src/include/tlvf/WSC/configData.h [hunk 3]
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
...
+
+        // Skip Vendor Extension attribute if it is not a WFA Vendor Extension attribute
+        auto vendor_ext_attr = getAttr<cWscAttrVendorExtension>();
+        if ((eWscVendorId::WSC_VENDOR_ID_WFA_1 != vendor_ext_attr->vendor_id_0()) ||
+            (eWscVendorId::WSC_VENDOR_ID_WFA_2 != vendor_ext_attr->vendor_id_1()) ||
+            (eWscVendorId::WSC_VENDOR_ID_WFA_3 != vendor_ext_attr->vendor_id_2())) {
+            return default_value;
```
Comment (arnout) [inline framework/tlvf/src/include/tlvf/WSC/configData.h]: To be fully correct, we should actually iterate over all vendor extension attributes until we find one that matches the WFA ID.

File: framework/tlvf/src/include/tlvf/WSC/configData.h [hunk 4]
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
...
+        }
+
+        size_t index     = 0;
+        auto vendor_data = vendor_ext_attr->vendor_data();
+        while ((index + sizeof(sWscWfaVendorExtSubelementMultiApIdentifier)) <=
+               vendor_ext_attr->vendor_data_length()) {
+            if (eWscWfaVendorExtSubelement::MULTI_AP_IDENTIFIER == vendor_data[index++]) {
```
Comment (arnout) [inline framework/tlvf/src/include/tlvf/WSC/configData.h]: (nitpick) I heavily dislike the use of ++ like this. I prefer explicit indexing, i.e. keep index constant within one iteration and update it at the end of the loop body only.

Also, I prefer explicit variables (usually I don't like adding variables like that, but here I think it documents the meaning of the individual bytes):
```
auto subelement_type = vendor_data[index];
auto subelement_length = vendor_data[index + 1];
if (subelement_type == ... && subelement_length == 1) {
    return static_cast<WSC::eWscVendorExtSubelementBssType>(vendor_data[index + 2]);
}
```

File: framework/tlvf/src/include/tlvf/WSC/configData.h [hunk 5]
```diff
@@ -60,7 +60,36 @@ class configData : public AttrList {
...
+                if (0x01 != vendor_data[index++]) {
+                    return default_value;
+                }
+
+                return static_cast<WSC::eWscVendorExtSubelementBssType>(vendor_data[index]);
+            }
+
+            // Skip this subelement
+            index += vendor_data[index++];
```
Comment (arnout) [inline framework/tlvf/src/include/tlvf/WSC/configData.h]: So this is why I don't like `index++`... It *looks* like index is incremented 3 times at first sight, you have to really follow the flow to understand that this is `vendor_data[index + 1]`...
```suggestion
            index += 2 + subelement_length;
```

File: framework/tlvf/src/src/WSC/m1.cpp [hunk 6]
```diff
@@ -204,21 +204,24 @@ bool m1::init(const config &cfg)
...
+    auto vendor_data = vendor_ext_attr->vendor_data();
 
-    sWscAttrVersion2 version2;
-    version2.struct_swap();
-    uint8_t *version2_buf = reinterpret_cast<uint8_t *>(&version2);
-    std::copy(version2_buf, version2_buf + sizeof(version2), vendor_ext_attr->vs_data());
+    // WFA Vendor Extension Subelement at #0: Version2
+    sWscWfaVendorExtSubelementVersion2 version2{eWscWfaVendorExtSubelement::VERSION2, 0x01,
+                                                eWscVendorExtVersionIE::WSC_VERSION2};
+    std::copy_n(reinterpret_cast<uint8_t *>(&version2), sizeof(version2), vendor_data);
```
Comment (arnout) [inline framework/tlvf/src/src/WSC/m1.cpp:224]: Since this is used both in M1 and M2, it could be worth factoring it in a function (nitpick).

**Review by arnout — APPROVED** [review]
Feel free to ignore all of my comments (though I really would like to fix that index++ stuff, I had to read it four times before I was certain the code was indeed correct...)

File: framework/tlvf/src/include/tlvf/WSC/configData.h (hunk 2 above)
Comment (mariomaz) [reply framework/tlvf/src/include/tlvf/WSC/configData.h]: Done

File: framework/tlvf/src/include/tlvf/WSC/configData.h (hunk 4 above)
Comment (mariomaz) [reply framework/tlvf/src/include/tlvf/WSC/configData.h]: Done

File: framework/tlvf/src/include/tlvf/WSC/configData.h (hunk 5 above)
Comment (mariomaz) [reply framework/tlvf/src/include/tlvf/WSC/configData.h]: Done

File: framework/tlvf/src/include/tlvf/WSC/configData.h (hunk 3 above)
Comment (mariomaz) [reply framework/tlvf/src/include/tlvf/WSC/configData.h]: Done

**Pull request merged**